pkg/streamingpromql/operators/functions/histograms.go

```diff
@@ -0,0 +1,410 @@
+	for g.remainingSeriesCount > 0 {
+		s, err := h.inner.NextSeries(ctx)
+		if err != nil {
+			if errors.Is(err, types.EOS) {
+				return fmt.Errorf("exhausted series before all groups were completed: %w", err)
+			}
+			return err
+		}
+		thisSeriesGroups := h.seriesGroupPairs[h.currentInnerSeriesIndex]
+
+		// Native histograms only ever go to their original labelset.
+		// Floats only ever go to their group.
+		// It is possible that a series has both floats and histograms.
+		// It is also possible that both series groups are the same.
+		// The conflict in points is then detected in computeOutputSeriesForGroup.
+		h.saveNativeHistogramsToGroup(s.Histograms, thisSeriesGroups.nativeHistogramGroup)
+		err = h.saveFloatsToGroup(s.Floats, thisSeriesGroups.bucketValue, thisSeriesGroups.classicHistogramGroup)
+		if err != nil {
+			return err
+		}
+
+		// We are done with the fPoints, so return these now
+		// hPoint's are returned to the pool after computeOutputSeriesForGroup is finished with them as they may be copied to a group.
+		types.FPointSlicePool.Put(s.Floats, h.memoryConsumptionTracker)
+		h.currentInnerSeriesIndex++
+	}
+	return nil
+}
+
+// saveFloatsToGroup places each fPoint into a bucket with the upperBound set by the input series.
+func (h *HistogramFunctionOverInstantVector) saveFloatsToGroup(fPoints []promql.FPoint, le string, g *bucketGroup) error {
+	if len(fPoints) > 0 {
+		upperBound, err := strconv.ParseFloat(le, 64)
+		if err != nil {
+			// The le label was invalid. Record it:
+			h.annotations.Add(annotations.NewBadBucketLabelWarning(
+				h.innerSeriesMetricNames.GetMetricNameForSeries(h.currentInnerSeriesIndex),
+				le,
+				h.inner.ExpressionPosition(),
+			))
+		} else {
+			if g.pointBuckets == nil {
+				g.pointBuckets, err = pointBucketPool.Get(h.timeRange.StepCount, h.memoryConsumptionTracker)
+				if err != nil {
+					return err
+				}
+				g.pointBuckets = g.pointBuckets[:h.timeRange.StepCount]
+			}
+			for _, f := range fPoints {
+				pointIdx := h.timeRange.PointIndex(f.T)
+				g.pointBuckets[pointIdx] = append(
+					g.pointBuckets[pointIdx],
+					bucket{
+						upperBound: upperBound,
+						count:      f.F,
+					},
+				)
+			}
+		}
+	}
+	g.remainingSeriesCount--
+	return nil
+}
+
+// saveNativeHistogramsToGroup stores the given native histograms onto the given group.
+// There should only ever be one native histogram per a series or series group. In general,
+// they are per-series, but we handle them as parts of groups because they could hypothetically
+// have the same series name as a classic histogram.
+func (h *HistogramFunctionOverInstantVector) saveNativeHistogramsToGroup(hPoints []promql.HPoint, g *bucketGroup) {
+	if len(hPoints) > 0 {
+		// We should only ever see one set of native histograms per group.
+		if g.nativeHistograms != nil {
+			panic("We should never see more than one native histogram per group")
```

Comment:
Rather than panicking and crashing the process, should we just return an error? 

If there's a situation that somehow does this, I don't think panicking is necessary given other queries can safely continue (contrast this with the case where a slice has been returned to the pool multiple times, where we want to crash as soon as possible to avoid returning corrupted query results).

Reply:
If this happens something has gone very wrong with the rest of the logic, so I wouldn't trust the results. We're more likely not notice a process panic'ing rather than just logging errors. Happy to change it though if you'd like.

Reply:
I think returning an error here is sufficient - this will still stop us returning a result for this query, and won't impact other inflight queries in the same process.